pecos/xmc/base.py

```diff
@@ -1411,6 +1484,117 @@ def predict(
 
             return pred_csr
 
+    def predict_select_outputs(
+        self,
+        X,
+        select_outputs_csr,
+        pred_params=None,
+        **kwargs,
+    ):
+        """Predict on given input data
+        Args:
+            X (csr_matrix or ndarray): instance feature matrix to predict on
+            select_outputs_csr (csr_matrix): the select outputs to predict
+            pred_params (HierarchicalMLModel.PredParams, optional): instance of HierarchicalMLModel.PredParams.
+                Default None to use the pred_params used in model training.
+            kwargs: overriding prediction parameters for backward compatibility
+                post_processor (str, optional):  override the post_processor specified in pred_params (all layers)
+                    Default None to disable overriding
+                threads (int, optional): the number of threads to use for training.
+                    Defaults to -1 to use all
+        Returns:
+            pred_csr (csr_matrix): prediction matrix (nr_inst, nr_labels)
+        """
+        if X.dtype != np.float32:
+            raise ValueError("X.dtype = {} is not supported".format(X.dtype))
+        if not isinstance(X, smat.csr_matrix) and not (
+            isinstance(X, np.ndarray) and X.flags["C_CONTIGUOUS"]
+        ):
+            raise ValueError("type(X) = {} is not supported".format(type(X)))
+        if X.shape[1] != self.nr_features:
+            raise ValueError("Feature dimension of query matrix does not match weight matrix")
+
+        if not isinstance(select_outputs_csr, smat.csr_matrix):
+            raise ValueError(
+                "type(select_outputs_csr) = {} is not supported".format(type(select_outputs_csr))
+            )
+        if select_outputs_csr.shape[1] != self.nr_labels:
+            raise ValueError("Label dimension of select output matrix does not match")
+
+        if X.shape[0] != select_outputs_csr.shape[0]:
+            raise ValueError("Instance dimension of query and select output matrix do not match")
+
+        # construct pred_params
+        if pred_params is None:
+            pred_params = self.get_pred_params()
+        elif isinstance(pred_params, self.PredParams):
+            pred_params = self.PredParams.from_dict(pred_params)
+            pred_params = self._duplicate_fields_with_name_ending_with_chain(
+                pred_params, self.PredParams, self.depth
+            )
+        else:
+            raise ValueError("unknown type(pred_params)!!")
+        pred_params.override_with_kwargs(kwargs)
+
+        if self.is_predict_only:
+            if self.weight_matrix_type != "CSC":
```

Comment:
Is it sufficient to check the bottom layer matrix type here or you need to check for all the layers?

Reply:
Change has been made to check the weight matrix type for every layer in a model to be "CSC" since theoretically a HierarchicalMLModel could support having layers with different matrix types.